// @shopify/app/dist/cli/commands/app/config/push.js
import { appFlags } from '../../../flags.js';
import Command from '../../../utilities/app-command.js';
import { pushConfig } from '../../../services/app/config/push.js';
import { load as loadApp } from '../../../models/app/loader.js';
import { getAppInfo } from '../../../services/local-storage.js';
import { Flags } from '@oclif/core';
import { globalFlags } from '@shopify/cli-kit/node/cli';
class ConfigPush extends Command {
    async run() {
        const { flags } = await this.parse(ConfigPush);
        const configName = flags.config || getAppInfo(flags.path)?.configFile;
        const app = await loadApp({ specifications: [], configName, directory: flags.path, mode: 'report' });
        await pushConfig({ app });
    }
}
ConfigPush.hidden = true;
ConfigPush.description = "Push your app's config to the Partner Dashboard.";
ConfigPush.flags = {
    ...globalFlags,
    ...appFlags,
    config: Flags.string({
        hidden: false,
        description: 'Name for the config file.',
        env: 'SHOPIFY_FLAG_APP_CONFIG',
    }),
};
export default ConfigPush;
